lib/gems/pending/util/object_storage/miq_s3_storage.rb

```diff
@@ -0,0 +1,101 @@
+require 'util/miq_object_storage'
+
+class MiqS3Storage < MiqObjectStorage
+  attr_reader :bucket_name
+
+  def self.uri_scheme
+    "s3".freeze
+  end
+
+  def self.new_with_opts(opts)
+    new(opts.slice(:uri, :username, :password, :region))
+  end
+
+  def initialize(settings)
+    super(settings)
+
+    # NOTE: This line to be removed once manageiq-ui-class region change implemented.
+    @settings[:region] ||= "us-east-1"
+    @bucket_name         = URI(@settings[:uri]).host
+
+    raise "username, password, and region are required values!" if @settings[:username].nil? || @settings[:password].nil? || @settings[:region].nil?
+  end
+
+  # Extract the path from the URI, so strip off the "s3://" scheme, the bucket
+  # hostname, leaving only the path minus the leading '/'
+  def uri_to_object_key(remote_file)
+    # `path` is `[5]` in the returned result of URI.split
+    URI.split(remote_file)[5][1..-1]
+  end
+
+  def upload_single(dest_uri)
+    object_key = uri_to_object_key(dest_uri)
+    logger.debug("Writing [#{source_input}] to => Bucket [#{bucket_name}] Key [#{dest_uri}]")
+
+    with_standard_s3_error_handling("uploading", source_input) do
+      bucket.object(object_key).upload_stream do |write_stream|
+        IO.copy_stream(source_input, write_stream, byte_count)
+      end
+    end
+  end
+
+  def download_single(source, destination)
+    object_key = uri_to_object_key(remote_file)
+    logger.debug("Downloading [#{source}] from bucket [#{bucket_name}] to local file [#{destination}]")
+
+    with_standard_s3_error_handling("downloading", source) do
+      if destination.kind_of?(IO)
+        s3.client.get_object(:bucket => bucket_name, :key => object_key) do |chunk|
+          destination.write(chunk)
+        end
+      else # assume file path
+        bucket.object(source).download_file(destination)
+      end
+    end
+    local_file
+  end
+
+  # no-op mostly
+  #
+  # dirs don't need to be created ahead of time in s3, unlike mounted file
+  # systems.
+  #
+  # For convenience though, calling bucket, which will initialize and create
+  # (if needed) the s3 bucket to be used for this instance.
+  def mkdir(_dir)
+    bucket
+  end
+
+  def bucket
+    @bucket ||= s3.bucket(bucket_name).tap do |bucket|
+      if bucket.exists?
+        logger.debug("Found bucket #{bucket_name}")
+      else
+        logger.debug("Bucket #{bucket_name} does not exist, creating.")
+        bucket.create
+      end
+    end
+  end
+
+  private
+
+  def s3
+    require 'aws-sdk'
+    require 'util/extensions/aws-sdk/s3_upload_stream_patch'
```

Comment:
Bah... this means that this didn't get loaded in the specs... which would have made things easier...

Oh well I guess, we caught the PR updating this, so I guess all is good.